text/3453-f16-and-f128.md

```diff
@@ -0,0 +1,124 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`f16` and `f128` are primitive floating types, they can be used like `f32` or `f64`. They always conform to binary16 and binary128 formats defined in the IEEE 754 standard, which means size of `f16` is always 16-bit, the size of `f128` is always 128-bit, the amount of exponent and mantissa bits follows the standard, and all operations are IEEE 754-compliant.
+
+```rust
+let val1 = 1.0; // Default type is still f64
+let val2: f128 = 1.0; // Explicit f128 type, casts from f64 literal
+let val3: f16 = 1.0; // Explicit f16 type, casts from f64 literal
+let val4 = 1.0f128; // Suffix of f128 literal
+let val5 = 1.0f16; // Suffix of f16 literal
+
+println!("Size of f128 in bytes: {}", std::mem::size_of_val(&val2)); // 16
+println!("Size of f16 in bytes: {}", std::mem::size_of_val(&val3)); // 2
+```
+
+Every target should support `f16` and `f128`, either in hardware or software. Most platforms do not have hardware support and therefore will need to use a software implementation.
+
+All [operators](https://doc.rust-lang.org/stable/std/primitive.f64.html#trait-implementations), [constants](https://doc.rust-lang.org/stable/std/f64/consts/), and [math functions](https://doc.rust-lang.org/stable/std/primitive.f64.html#implementations) defined for `f32` and `f64` in core, must also be defined for `f16` and `f128`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `f16` type
+
+`f16` consists of 1 bit of sign, 5 bits of exponent, 10 bits of mantissa. It is exactly equivalent to the 16-bit IEEE 754 binary16 [Half-precision floating-point format](https://en.wikipedia.org/wiki/Half-precision_floating-point_format).
+
+The following traits will be implemented for conversion between `f16` and other types:
+
+```rust
+impl From<f16> for f32 { /* ... */ }
+impl From<f16> for f64 { /* ... */ }
+impl From<bool> for f16 { /* ... */ }
+impl From<u8> for f16 { /* ... */ }
+impl From<i8> for f16 { /* ... */ }
+```
+
+`f16` will generate `half` type in LLVM IR.
+
+On the hardware level, `f16` can be accelerated on RISC-V via [the Zfh or Zfhmin extensions](https://five-embeddev.com/riscv-isa-manual/latest/zfh.html), on x86 with AVX-512 via [the FP16 instruction set](https://en.wikipedia.org/wiki/AVX-512#FP16), and on [some Arm platforms](https://developer.arm.com/documentation/100067/0607/Other-Compiler-specific-Features/Half-precision-floating-point-number-format). Most platforms do not have hardware support and therefore will need to use a software implementation.
```

Comment:
it's supported on PowerISA too, when the right optional features are implemented in HW.

Reply:
Can you provide a link to which optional feature this is? Also, it's not documented on the Wikipedia articles for [Power ISA](https://en.wikipedia.org/wiki/Power_ISA) or [Half-precision hardware support](https://en.wikipedia.org/wiki/Half-precision_floating-point_format#Hardware_support), please add this information to those articles too.

Reply:
`f16` is supported with VSX on PowerISA v3.1+ (mostly just conversions only -- see `xvcvhpsp` at [PowerISA v3.1B](https://files.openpower.foundation/s/dAYSdGzTfW4j2r2) page 858) or with [SVP64](https://libre-soc.org/openpower/sv/) (a draft PowerISA extension that [provides full HW support](https://git.libre-soc.org/?p=libreriscv.git;a=blob;f=openpower/sv/rfc/ls005.xlen.mdwn;hb=5e573680771f7a041d93d394003d6f9f08177a98#l131))